smtpserver/server.go

```diff
@@ -1899,6 +1900,34 @@ func (c *conn) submit(ctx context.Context, recvHdrFor func(string) string, msgWr
 		xsmtpUserErrorf(smtp.C550MailboxUnavail, smtp.SePol7DeliveryUnauth1, "must match authenticated user")
 	}
 
+	// Check if the message contains non-ascii characters. If no such characters are found,
+	// the SMTPUTF8 extension is not required.
+	// ../rfc/6531:497
+	isASCII := func(s string) bool {
+		for _, c := range s {
+			if c > unicode.MaxASCII {
+				return false
+			}
+		}
+		return true
+	}
+	c.smtputf8 = !isASCII(c.mailFrom.Localpart.String())
+	for _, rcpt := range c.recipients {
+		if !isASCII(rcpt.rcptTo.Localpart.String()) {
+			c.smtputf8 = true
+			break
+		}
+	}
+	for _, values := range header {
```

Comment:
i've read a bit more into rfc 653[012], which confirms that headers of multiparts can also have utf8, which is also a reason to use smtputf8. so this check is not exhaustive.

for example, see this message by arnt: https://github.com/arnt/eai-test-messages/blob/master/attachment
a smaller message like this would be a good test to add.

the easiest option is probably to just read all headers of all parts, and check for non-ascii. we would need a parsed message.Part. it is currently parsed in message.From() at the beginning of cmdSubmit, but we probably need to refactor a little bit to parse the part earlier, and perhaps just pass it to message.From (we don't want to parse messages multiple times).

something like this will do the check:

```
	hasNonASCII := func(r io.Reader) bool {
		br := bufio.NewReader(r)
		for {
			b, err := br.ReadByte()
			if err == io.EOF {
				break
			}
			xcheckf(err, "read header")
			if b > 0x7f {
				return true
			}
		}
		return false
	}

	var walk func(p *message.Part) bool
	walk = func(p *message.Part) bool {
		if hasNonASCII(p.HeaderReader()) {
			return true
		}
		for _, pp := range p.Parts {
			if walk(&pp) {
				return true
			}
		}
		return false
	}

	fmt.Println("message needs smtputf8:", walk(&part))
```

Reply:
Done.

But I'm not sure the way I did it to avoid parsing the message twice is the best choice.

Reply:
looks good to me!